runtime/jcl/common/sun_misc_Unsafe.cpp

```diff
@@ -871,12 +934,139 @@ Java_jdk_internal_misc_Unsafe_registerNatives(JNIEnv *env, jclass clazz)
+		if (sourceOffset != destOffset) {
+			return JNI_TRUE;
+		}
+	}
+	return JNI_FALSE;
+}
+
 void JNICALL
-Java_jdk_internal_misc_Unsafe_copySwapMemory0(JNIEnv *env, jobject receiver, jobject obj1, jlong size1, jobject obj2, jlong size2, jlong size3, jlong size4)
+Java_jdk_internal_misc_Unsafe_copySwapMemory0(JNIEnv *env, jobject receiver, jobject srcBase, jlong srcOffset,
+		jobject dstBase, jlong dstOffset, jlong copySize, jlong elemSize)
 {
-	/* stub implementation */
-	/* private native void copySwapMemory0(java.lang.Object, long, java.lang.Object, long, long, long); */
-	assert(!"Java_jdk_internal_misc_Unsafe_copySwapMemory0 is unimplemented");
+	J9VMThread *currentThread = (J9VMThread*)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	UDATA sourceOffset = (UDATA)srcOffset;
+	UDATA destOffset = (UDATA)dstOffset;
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+	UDATA actualCopySize = (UDATA)copySize;
+	UDATA actualElementSize = (UDATA)elemSize;
+
+	if ((copySize < 0) || (copySize != (jlong)actualCopySize) || (elemSize != (jlong)actualElementSize)) {
+illegal:
+		vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, NULL);
+	} else if ( ! (2 == elemSize || 4 == elemSize || 8 == elemSize)) {
+		/* verify that element size is supported for swapping */
+		goto illegal;
+	} else if (0 != (copySize % elemSize)) {
+		/* verify that size to be copied is a multiple of element size */
+		goto illegal;
+	} else {
+		j9object_t sourceObject = NULL;
+		j9object_t destObject = NULL;
+		if (NULL != srcBase) {
+			sourceObject = J9_JNI_UNWRAP_REFERENCE(srcBase);
+			J9Class *clazz = J9OBJECT_CLAZZ(currentThread, sourceObject);
+			if (!J9CLASS_IS_ARRAY(clazz)) {
+				goto illegal;
+			}
+			if (!J9ROMCLASS_IS_PRIMITIVE_TYPE(((J9ArrayClass*)clazz)->componentType->romClass)) {
+				goto illegal;
+			}
+		}
+		if (NULL != dstBase) {
+			destObject = J9_JNI_UNWRAP_REFERENCE(dstBase);
+			J9Class *clazz = J9OBJECT_CLAZZ(currentThread, destObject);
+			if (!J9CLASS_IS_ARRAY(clazz)) {
+				goto illegal;
+			}
+			if (!J9ROMCLASS_IS_PRIMITIVE_TYPE(((J9ArrayClass*)clazz)->componentType->romClass)) {
+				goto illegal;
+			}
+		}
+
+		jboolean overlapIsUnaligned = memOverlapIsUnaligned(sourceObject, sourceOffset, destObject, destOffset, actualCopySize);
+		jboolean overlapIsNone = memOverlapIsNone(sourceObject, sourceOffset, destObject, destOffset, actualCopySize);
+
+		/* copy source data to destination first if source and destination memory is overlapping
+		 * and the overlap is unaligned. This will prevent any errors during swapping.
+		 */
+		if (overlapIsUnaligned) {
+			copyMemory(currentThread, sourceObject, sourceOffset, destObject, destOffset, actualCopySize);
+		}
+
+		/* use temporary byte for swap if memory is overlapping */
+		UDATA tempOffset = 0;
+		if (! overlapIsNone) {
+			tempOffset = (UDATA)unsafeAllocateMemory(currentThread, (UDATA)1, FALSE);
```

Comment:
Why malloc this?  Could a stack allocated value (local variable) be used instead?